### ISSUE TITLE
Is there a technical reason why some of the header definitions are "static inline" and so don't have exported symbols?

### ISSUE THREAD

**Issue by GavinRay97** [axboe/liburing#660]
I've been trying to integrate `liburing` with the JVM's new Panama functionality, which allows you to do FFI and interop with foreign memory without writing bindings/stubs.

The major issue I've run into is that the majority of the API seems to be defined as `static inline`, so when I try to call them, they simply don't exist:

- https://github.com/GavinRay97/panama-liburing/blob/82f83a247bb78c26d787992fca28ea1b951dfd3c/lib/src/test/java/panama/liburing/LibraryTest.java#L38-L44

```java
SegmentAllocator allocator = SegmentAllocator.newNativeArena(session);

// Works
MemorySegment ring = io.uring.io_uring.allocate(allocator);
io_uring_queue_init(QUEUE_DEPTH, ring, 0);

// Works
MemoryAddress sqe = io_uring_get_sqe(ring);

// ERROR HERE:
//
// java.lang.UnsatisfiedLinkError: unresolved symbol: io_uring_prep_rw
io_uring_prep_rw(/* op */ io.uring.liburing_h.IORING_OP_NOP(),
    /* sqe */ sqe, /* fd */ 0, /* addr */ MemoryAddress.NULL, /* len */ 0, /* offset */ 0);
```

I don't know much about C/systems programming so I thought there might be a good reason why they were like this.

If the changes can't be made repo-wide, is there any detriment to applying a patch on top of a personal library that removes the `static inline` from the definitions?

Thank you =)

**Comment by axboe** [comment]
If it's in a header, it needs to be static inline or you'll end up with duplicates. And the reason they are static inline rather than an exported symbol is to avoid the function call overhead for them.

For a private library, you can move them into C files for example and just export them. It'll still work fine, no functionality depends on them being inline.

**Comment by notinaboat** [comment]
@axboe I ran into the same issue calling `liburing` from Julia.

I am guessing that the same issue will effect each new language that wants to interface with `liburing`.

Would you consider adding a C file to the library with functions something like `_noinline__io_uring_prep_poll_add()` to wrap each of the `static inline` functions ?

The alternatives are:
 1. Each language interface library builds its own C wrapper for the `static inline` functions
 2. Each language interface library translates the content of the `static inline` functions into the target language.

Either of these alternatives increase the risk of errors due to repetition of work.
It would be really great to be able to talk to `liburing` directly from non-C languages without relying on a hand-coded shim, or patch set that ends up being out of date.

**Comment by axboe** [comment]
@notinaboat Yes, I'd be open to creating such a wrapper as long as it doesn't get in the way of C usage obviously. I agree that it'd be nice to have it in liburing, rather than have each non-C like language create their own.

Would you be up for creating a PR for that?
